In [1]:
from utils import *
import carla
from detectors import torch_script_module
from carla_control import carla_client
from utils.mods.post_processing import mono_label_distance_tracker, multi_classes_assemble_tracker
import numpy as np
import open3d as o3d
import torch
import time

from IPython.display import clear_output, display

Jupyter environment detected. Enabling Open3D WebVisualizer.
[Open3D INFO] WebRTC GUI backend enabled.
[Open3D INFO] WebRTCWindowSystem: HTTP handshake server disabled.


In [2]:
logger = create_logger()

In [3]:
# client = carla_client(host = "192.168.31.13")
client = carla_client()

In [4]:
client.init_client()
client.start_client()

In [5]:
client.debug_luanch_test()

CFG_FILE = ".\\configs\\kitti_prepocessing.yaml"
cfg = cfg_from_yaml_file(CFG_FILE, cfg)

pcs_dataset = carla_point_cloud_dataset(dataset_cfg = cfg.DATA_CONFIG, logger=logger, lidar=client.lidar_t, class_names=cfg.CLASS_NAMES)
object_tracker = multi_classes_assemble_tracker(num_classes=4, track_length=25, multi_head=True)
# model = torch_script_module(path="./torch_scripts/point_pillar_model.pt")
# model.load_model()
model = torch.jit.load("./torch_scripts/point_pillar_model.pt")
model.cuda()
model.eval()
print(cfg)

{'ROOT_DIR': WindowsPath('D:/Python_project/beamng_python_experiment'), 'LOCAL_RANK': 0, 'CLASS_NAMES': ['Car', 'Pedestrian', 'Cyclist'], 'DATA_CONFIG': {'POINT_CLOUD_RANGE': [0, -39.68, -3, 69.12, 39.68, 1], 'POINT_FEATURE_ENCODING': {'encoding_type': 'absolute_coordinates_encoding', 'used_feature_list': ['x', 'y', 'z', 'intensity'], 'src_feature_list': ['x', 'y', 'z', 'intensity']}, 'DATA_PROCESSOR': [{'NAME': 'utils.mods.pre_pocessing.robo3d.simulate_fog', 'EXT_MODULE': True, 'GAIN': False, 'INTE_NAME': 'integral_lookup_tables_seg_light_0.008beta', 'NOISE': 10, 'BETA': 0.2, 'NOISE_VARIANT': 'V1'}, {'NAME': 'transform_points_to_voxels', 'VOXEL_SIZE': [0.16, 0.16, 4], 'MAX_POINTS_PER_VOXEL': 32, 'MAX_NUMBER_OF_VOXELS': {'train': 16000, 'test': 40000}}]}, 'MODEL': {'NAME': 'PointPillar', 'POST_PROCESSING': {'RECALL_THRESH_LIST': [0.3, 0.5, 0.7], 'SCORE_THRESH': 0.1, 'OUTPUT_RAW_SCORE': False, 'EVAL_METRIC': 'kitti', 'NMS_CONFIG': {'MULTI_CLASSES_NMS': False, 'NMS_TYPE': 'nms_gpu', 'NMS

In [6]:
def carla_ticking(idx, data_dict): 
    before_time = time.perf_counter()
    
    client.carla_world.tick()
    spectator = client.carla_world.get_spectator()
    transform = client.vehicle.get_transform()
    spectator.set_transform(carla.Transform(transform.location + carla.Location(z=30),
                                                carla.Rotation(pitch=-90)))
    after_time = time.perf_counter()
    return after_time - before_time, None
    
def model_forwarding(idx, data_dict):
    final_boxes, final_scores, final_labels = None, None, None
    before_time = time.perf_counter()
    if data_dict["points"] is not None:
        data_dict = pcs_dataset.collate_batch([data_dict])
        load_data_to_gpu(data_dict)

        pred_dicts, _ = model.forward(data_dict)
        
        cls_preds = pred_dicts["pred_scores"]
        box_preds = pred_dicts["pred_boxes"]
        label_preds = pred_dicts["pred_labels"]

        selected, selected_scores = class_agnostic_nms(
                            box_scores=cls_preds, box_preds=box_preds,
                            score_thresh=0.4
                        )

        final_scores = selected_scores
        final_labels = label_preds[selected]
        final_boxes = box_preds[selected]
    after_time = time.perf_counter()
    return after_time - before_time, data_dict, final_boxes, final_scores, final_labels

def scene_rendering(idx, data_dict, vis, final_boxes, final_scores, final_labels):
    before_time = time.perf_counter()
    if data_dict["points"] is not None:
        draw_scenes(vis,
                    points=data_dict['points'][:, 1:], ref_boxes=final_boxes,
                    ref_scores=final_scores, ref_labels=final_labels, confidence=None
                )

    vis.poll_events()
    vis.update_renderer()
    vis.clear_geometries()
    after_time = time.perf_counter()
    return after_time - before_time, None

In [7]:
vis = o3d.visualization.Visualizer()
vis.create_window()

with torch.no_grad():
    for idx, data_dict in enumerate(pcs_dataset):

        pre_time = data_dict["pre_time"]
        
        ticking_time, _ = carla_ticking(idx, data_dict)
        
        if data_dict["points"] is not None:
        
            forward_time, data_dict, final_boxes, final_scores, final_labels = model_forwarding(idx, data_dict)

            render_time, _ = scene_rendering(idx, data_dict, vis, final_boxes, final_scores, final_labels)
            
            render_time = time.perf_counter()
            
            # if final_boxes is not None:
            #     object_tracker.updates_object(final_boxes[np.newaxis, :, :], final_labels[np.newaxis, :], final_scores[np.newaxis, :])

            # bounding_boxes, box_scores, box_labels, tracks = object_tracker.get_all_object()

            # draw_scenes(vis,
            #     points=data_dict['points'][:, 1:], ref_boxes=bounding_boxes,
            #     ref_scores=box_scores, ref_labels=box_labels, confidence=None, tracks=tracks
            # )
            # draw_scenes(vis,
            #     points=data_dict['points'][:, 1:], ref_boxes=None,
            #     ref_scores=None, ref_labels=None, confidence=None, tracks=None
            # )
            # vis.poll_events()
            # vis.update_renderer()
            # vis.clear_geometries()
            
            render_time = time.perf_counter() - render_time
        
            clear_output(wait=True)
            logger.info(f"Compute time: {pre_time:.3f} + {ticking_time:.3f} + {forward_time:.3f} + {render_time:.3f} == {pre_time + ticking_time + forward_time + render_time:.3f}s")
            logger.info(f"Target amount: {len(final_boxes if (final_boxes is not None) else [])}")
            logger.info(f"current uuid:{object_tracker.get_last_uuid()}")
            


2023-10-05 11:45:19,480   INFO  Compute time: 0.462 + 0.043 + 0.033 + 0.000 == 0.538s
2023-10-05 11:45:19,480   INFO  Target amount: 12
2023-10-05 11:45:19,480   INFO  current uuid:[0, 0, 0, 0]


(124788, 4)
(57300, 4)

{
	"class_name" : "ViewTrajectory",
	"interval" : 29,
	"is_loop" : false,
	"trajectory" : 
	[
		{
			"boundingbox_max" : [ 69.118263244628906, 39.679920196533203, 16.415634155273438 ],
			"boundingbox_min" : [ -0.059999999999999998, -39.679874420166016, -6.9146575927734375 ],
			"field_of_view" : 60.0,
			"front" : [ -0.90307097537632919, 0.0017988087570628851, 0.42948757574567964 ],
			"lookat" : [ 34.529131622314452, 2.288818359375e-05, 4.75048828125 ],
			"up" : [ 0.42948904059539766, 0.0070563614983622357, 0.90304450154510629 ],
			"zoom" : 0.69999999999999996
		}
	],
	"version_major" : 1,
	"version_minor" : 0
}